### PR TITLE
feat: auto increment support

### PR DESCRIPTION
**Pull request by p5quared**
#### Description of changes

Transformer changes to support auto increment (`serial`) fields on Postgres datasources. Accomplished by making the 'create' input types optional when `@default` is applied.

1. Definition of `@default` has changed, its `value: String` parameter is now optional.
2. `default-value-transformer` allows no `value` on Int fields if the datasource is Postgres
3. Unit tests for the above

##### CDK / CloudFormation Parameters Changed

n/a

#### Description of how you validated changes

Unit tests, _existing_ e2e tests, `testTransform`. Will write e2e test after implementing changes to Lambda layer.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)

### Codebuild Tests
Ran against beta via `yarn split-e2e-tests-beta`.

- [Specific e2e](https://us-east-1.console.aws.amazon.com/codesuite/codebuild/594813022831/projects/amplify-category-api-e2e-workflow/batch/amplify-category-api-e2e-workflow:c9868387-2099-4a8d-abf9-15a61c73603e?region=us-east-1)
- [cloud-e2e-beta](https://us-east-1.console.aws.amazon.com/codesuite/codebuild/455172627829/projects/amplify-category-api-e2e-workflow/batch/amplify-category-api-e2e-workflow%3Ac55e8e65-a6ea-41b7-95fd-13262257d5f4?region=us-east-1)
- [cloud-e2e-beta (latest)](https://us-east-1.console.aws.amazon.com/codesuite/codebuild/455172627829/projects/amplify-category-api-e2e-workflow/batch/amplify-category-api-e2e-workflow:ea24878f-7377-4e03-85c2-c711e7e01192?region=us-east-1)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts [hunk 1]
```diff
@@ -41,180 +46,35 @@ describe('DefaultValueModelTransformer:', () => {
     ).toThrow('The @default directive may only be added to scalar or enum field types.');
   });
 
-  it('throws if @default is used with a null value', () => {
+  it.each([
```
Comment (p5quared) [inline packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:49]: This change can be dropped. I just felt it helped me scrolling around through this file a bunch.

**Review by sundersc — APPROVED** [review]

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        ]),
+      );
+
+      // Check invalid CRUD operation returns generic error message
+      const createTodo6 = await coffeeQueueTableCRUDLHelper.create({ order: 'mocha' });
+
+      try {
+        await coffeeQueueTableCRUDLHelper.create({ orderNumber: createTodo6.orderNumber, order: 'americano' });
+      } catch (error) {
+        console.log('throw in create');
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:148]: remove

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        coffeeQueueTableCRUDLHelper.checkGenericError(error?.message);
+      }
+
+      const invalidOrderNumber = 99999999;
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:152]: I assume this is invalid because it exceeds the maximum value of a `SERIAL` field? If so, worth noting that in a comment, in case we start supporting PG versions that have a higher limit.

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+      const client = new LambdaClient({ region });
+      const functionName = outputs[name].SQLFunctionName;
+      const command = new GetProvisionedConcurrencyConfigCommand({
+        FunctionName: functionName,
+        Qualifier: resourceNames.sqlLambdaAliasName,
+      });
+      const response = await client.send(command);
+      expect(response.RequestedProvisionedConcurrentExecutions).toEqual(2);
+    });
+  });
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:186]: remove -- unrelated to this test

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        coffeeQueueTableCRUDLHelper.checkGenericError(error?.message);
+      }
+
+      try {
+        await coffeeQueueTableCRUDLHelper.delete({ orderNumber: invalidOrderNumber });
+      } catch (error) {
+        console.log('throw in delete');
+        coffeeQueueTableCRUDLHelper.checkGenericError(error?.message);
+      }
+    });
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:174]: Does PG allow someone to specify a value for a `SERIAL` field? If so, we need a test for that case. If not, we need a test that a customer cannot supply a value for it (which might just be a unit test that asserts the serial default field is not present in the input shape.

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts (hunk 1 above)
Comment (palpatim) [reply packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:49]: Not following your comment. This seems like a good refactor, so why should it be dropped?

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts [hunk 6]
```diff
@@ -338,4 +198,90 @@ describe('DefaultValueModelTransformer:', () => {
...
+    { typeStr: 'Boolean' },
+    { typeStr: 'AWSJSON' },
+    { typeStr: 'AWSDate' },
+    { typeStr: 'AWSDateTime' },
+    { typeStr: 'AWSTime' },
+    { typeStr: 'AWSTime' },
+    { typeStr: 'AWSURL' },
+    { typeStr: 'AWSPhone' },
+    { typeStr: 'AWSIPAddress' },
+  ])('throws if auto-increment is implied on non-int types', ({ typeStr }) => {
```
Comment (palpatim) [inline packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:231]: Can you confirm that the test output includes the `typeStr` value in the test log? If a test fails, we want to know which value is problematic.

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts [hunk 7]
```diff
@@ -338,4 +198,90 @@ describe('DefaultValueModelTransformer:', () => {
     expect(out.resolvers['Mutation.createNote.init.1.req.vtl']).toBeDefined();
     expect(out.resolvers['Mutation.createNote.init.2.req.vtl']).toBeUndefined();
   });
+
+  it.each([{ strategy: mockSqlDataSourceStrategy() }])(
```
Comment (palpatim) [inline packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:202]: Since it's important that this is a mysql data source, go ahead and specify the `dbType` argument for `mockSqlDataSourceStrategy`, so it's 1. clear that mysql is important; and 2. future-proof against default changes. (Which we hopefully won't do, but you never know...)

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts [hunk 8]
```diff
@@ -338,4 +198,90 @@ describe('DefaultValueModelTransformer:', () => {
...
+    `;
+      testTransform({
+        schema,
+        transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
+        dataSourceStrategies: constructDataSourceStrategies(schema, strategy),
+      });
+    }).toThrow('The empty @default (auto-increment) may only be applied to integer fields.');
+  });
+
+  it('should successfully transform a schema that implies auto-increment', async () => {
```
Comment (palpatim) [inline packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:249]: What does "implies auto-increment" mean?

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts [hunk 9]
```diff
@@ -338,4 +198,90 @@ describe('DefaultValueModelTransformer:', () => {
...
+        id: ID! @primaryKey
+        value: ${typeStr} @default
+      }
+    `;
+      testTransform({
+        schema,
+        transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
+        dataSourceStrategies: constructDataSourceStrategies(schema, strategy),
+      });
+    }).toThrow('The empty @default (auto-increment) may only be applied to integer fields.');
```
Comment (palpatim) [inline packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:246]: Error message nit: 
```suggestion
    }).toThrow('A @default directive with no arguments may only be applied to integer fields.');
```

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts [hunk 10]
```diff
@@ -338,4 +198,90 @@ describe('DefaultValueModelTransformer:', () => {
...
+        orderNumber: Int! @default
+        name: String
+      }`;
+      expect(() => {
+        testTransform({
+          schema: schema,
+          transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
+          dataSourceStrategies: constructDataSourceStrategies(schema, strategy),
+        });
+      }).toThrow('The @default directive requires a value property on non Postgres datasources.');
```
Comment (palpatim) [inline packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:217]: 1. Can you confirm that a `@default(value: ...)` works as expected on a mysql data source? (That is, it passes specified value in the input if the mutation input doesn't include it; and it allows overrides of the default if it IS specified). Best way to confirm that would be a test (possibly an already existing one!)
2. Do we have a test (possibly an already existing one!) to assert that `@default(value:...)` works for postgres?

File: packages/amplify-graphql-default-value-transformer/src/graphql-default-value-transformer.ts [hunk 11]
```diff
@@ -31,16 +31,24 @@ import { TypeValidators } from './validators';
 
 const nonStringTypes = ['Int', 'Float', 'Boolean', 'AWSTimestamp', 'AWSJSON'];
 
-const validateFieldType = (ctx: TransformerSchemaVisitStepContextProvider, type: TypeNode): void => {
+const validateFieldType = (ctx: TransformerSchemaVisitStepContextProvider, config: DefaultValueDirectiveConfiguration): void => {
+  const type = config.field.type;
+  const argc = config.directive.arguments!.length;
```
Comment (palpatim) [inline packages/amplify-graphql-default-value-transformer/src/graphql-default-value-transformer.ts:36]: Is `arguments` guaranteed to be non-null? If you can't guarantee it, it's safer to say 

```suggestion
  const argc = config.directive.arguments?.length ?? 0;
```

File: packages/amplify-graphql-default-value-transformer/src/graphql-default-value-transformer.ts [hunk 12]
```diff
@@ -31,16 +31,24 @@ import { TypeValidators } from './validators';
...
   const enums = ctx.output.getTypeDefinitionsOfKind(Kind.ENUM_TYPE_DEFINITION) as EnumTypeDefinitionNode[];
   if (isListType(type) || !isScalarOrEnum(type, enums)) {
     throw new InvalidDirectiveError('The @default directive may only be added to scalar or enum field types.');
   }
+  if (isPostgresModel(ctx, config.object.name.value) && argc === 0 && getBaseType(type) !== 'Int') {
+    throw new InvalidDirectiveError('The empty @default (auto-increment) may only be applied to integer fields.');
```
Comment (palpatim) [inline packages/amplify-graphql-default-value-transformer/src/graphql-default-value-transformer.ts:42]: see note above re error message. Also, maybe say "fields with Int type" or similar to make it clear that you're talking about the GraphQL type `Int`.

File: packages/amplify-graphql-transformer-core/src/index.ts [hunk 13]
```diff
@@ -61,6 +61,8 @@ export {
   isDefaultDynamoDbModelDataSourceStrategy,
   isDynamoDbModel,
   isDynamoDbType,
+  isPostgresModel,
+  isPostgresDbType,
```
Comment (palpatim) [inline packages/amplify-graphql-transformer-core/src/index.ts:65]: super nit: Please sort this list :)

File: packages/amplify-graphql-transformer-core/src/utils/model-datasource-strategy-utils.ts [hunk 14]
```diff
@@ -141,6 +141,27 @@ export const isSqlModel = (ctx: DataSourceStrategiesProvider, typename: string):
...
+    return false;
+  }
+  const modelDataSourceType = getModelDataSourceStrategy(ctx, typename);
+  return isPostgresDbType(modelDataSourceType.dbType);
+};
+
+/**
+ * Type predicate that returns true if `dbType` is a PostgreSQL database type
+ */
+export const isPostgresDbType = (dbType: ModelDataSourceStrategyDbType): dbType is ModelDataSourceStrategySqlDbType => {
```
Comment (palpatim) [inline packages/amplify-graphql-transformer-core/src/utils/model-datasource-strategy-utils.ts:161]: The return value of this function is misleading. You already know that `dbType` is a `ModelDataSourceStrategyDbType` because it's specified in the argument type. What the predicate actually does is assert that dbType is exactly equal to `POSTGRES`. Change the type predicate to assert the constant string:
```suggestion
export const isPostgresDbType = (dbType: ModelDataSourceStrategyDbType): dbType is 'POSTGRES' => {
```

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts (hunk 3 above)
Comment (p5quared) [reply packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:152]: Actually it is invalid because it is greater than any existing record's `orderNumber`. That purpose is definitely not clear since even the number 4 _should_ be invalid. I wrote some comments but then opted to use a more descriptive variable name:
```
const biggerThanAnyExistingOrderNumber = 99999999;
```

File: packages/amplify-graphql-default-value-transformer/src/graphql-default-value-transformer.ts (hunk 11 above)
Comment (p5quared) [reply packages/amplify-graphql-default-value-transformer/src/graphql-default-value-transformer.ts:36]: Applied changes

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    dbController: SqlDatatabaseController;
+    resourceNames: { sqlLambdaAliasName: string };
+  },
+  testBlockDescription: string,
+  engine: ImportedRDSType,
+): void => {
+  describe(`${testBlockDescription} - ${engine}`, () => {
+    const amplifyGraphqlSchema = `
+      type CoffeeQueue @model @refersTo(name: "${getRDSTableNamePrefix()}coffee_queue") {
+        orderNumber: Int! @primaryKey @default
```
[the remark this answers is not in the document]
Comment (p5quared) [reply packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:27]: Noted in comment

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts (hunk 5 above)
Comment (p5quared) [reply packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:174]: PG does allow users to specify a value for `SERIAL` fields (but does not stop users' sequences from producing that value later). I modified the test to use a custom value for one of the inserts.

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts (hunk 6 above)
Comment (p5quared) [reply packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts]: added `typeStr` to the logged message:
```
'throws if @default (no args) is used on non-int types (case: $typeStr)'
```

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts (hunk 7 above)
Comment (p5quared) [reply packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts]: Those points make a lot of sense. Added the argument.

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts (hunk 8 above)
Comment (p5quared) [reply packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts]: the only defined use of `@default` is for auto-increment fields but that is not explicit in the word `@default`. That being said I agree that my wording is clunky. Re-written.

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts (hunk 10 above)
Comment (p5quared) [reply packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts]: The only e2e usage of `@default` I could find only used DDB. I modified the general sql models test to utilize `@default(value)` which should verify behavior on both sources.

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+        order: String!
+        customer: String
+      }
+    `;
+
+    const { projFolderName, region, connectionConfigName, dbController, resourceNames } = options;
+    const templatePath = path.resolve(path.join(__dirname, '..', '__tests__', 'backends', 'sql-models'));
```
[the remark this answers is not in the document]
Comment (p5quared) [reply packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:35]: If it's okay I'd like to defer this migration to a separate PR and migrate a number of tests at once.

File: packages/amplify-graphql-api-construct-tests/src/__tests__/sql-pg-auto-increment.test.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    constructTestOptions('ssm'),
+    'creates a GraphQL API from SQL-based models with SSM Credential Store',
+    ImportedRDSType.POSTGRESQL,
+  );
+
+  testGraphQLAPIAutoIncrement(
+    constructTestOptions('connectionUri'),
+    'creates a GraphQL API from SQL-based models using Connection String SSM parameter',
+    ImportedRDSType.POSTGRESQL,
+  );
```
[the remark this answers is not in the document]
Comment (p5quared) [reply packages/amplify-graphql-api-construct-tests/src/__tests__/sql-pg-auto-increment.test.ts:60]: Per @phani-srikar's advice I've dropped `ssm`.

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (palpatim) [reply packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:35]: Fair enough, in fact let's defer it altogether, pending outcome of discussions per a similar convo in https://github.com/aws-amplify/amplify-category-api/pull/2906#discussion_r1784650482

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts (hunk 1 above)
Comment (palpatim) [reply packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:50]: Gotcha. It's a good change, thanks for adding it!

**Review by palpatim — DISMISSED** [review]

**Review by sundersc — APPROVED** [review]

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+        url: apiEndpoint,
+        region,
+        disableOffline: true,
+        auth: {
+          type: AUTH_TYPE.API_KEY,
+          apiKey,
+        },
+      });
+
+      coffeeQueueTableCRUDLHelper = new CRUDLHelper(appSyncClient, 'CoffeeQueue', 'CoffeeQueues', coffeeQueueFieldMap);
```
Comment (p5quared) [inline packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:55]: The only change since previous approval is the use of `coffeeQueueFieldMap` which is necessary after changes merged in #2906.

**Comment by p5quared** [comment]
Changes since approval:

1. Tiny update/refactor to E2E since #2906.
2. Ignore `types.ts` from `@default` test coverage and add a test case `it('cannot use `null` as a default value')` (on DynamoDB) to hit coverage requirements.

**Review by sundersc — APPROVED** [review]

File: packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts (hunk 18 above)
Comment (p5quared) [inline packages/amplify-graphql-api-construct-tests/src/sql-tests-common/sql-models-auto-increment.ts:55]: Since previous approval, we now use a `FieldMap` instead of a list of fields.

File: packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts [hunk 19]
```diff
@@ -338,4 +200,106 @@ describe('DefaultValueModelTransformer:', () => {
...
+        name: String
+      }`;
+
+    expect(() => {
+      testTransform({
+        schema: schema,
+        transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
+      });
+    }).toThrow('The @default directive does not support null values.');
+  });
```
Comment (p5quared) [inline packages/amplify-graphql-default-value-transformer/src/__tests__/amplify-grapphql-default-value-transformer.test.ts:218]: Also added this test case to hit coverage requirements.

**Review by palpatim — APPROVED** [review]
